lib/rubygems/test_case.rb

```diff
@@ -355,6 +355,13 @@ def setup
     Dir.chdir @tempdir
 
     ENV['HOME'] = @userhome
+    FileUtils.mkdir_p File.join(@userhome, ".gem")
+    File.write File.join(@userhome, ".gemrc"), "--- {}"
+    temp_cred = File.join(@userhome, '.gem', 'credentials')
+    FileUtils.mkdir_p File.dirname(temp_cred)
+    File.write temp_cred, ':rubygems_api_key: 701229f217cdf23b1344c7b4b54ca97'
+    File.chmod 0600, temp_cred
```

Comment:
In my opinion, only the tests that need this should create these files, and they should clean them up after themselves. That makes much more clear the side effects each test has, and also avoids hacks like the one you had to add to `test/rubygems/test_gem_commands_signout_command.rb` where you delete the files if they exists.

Reply:
It affects with the following tests.

* test_gem_config_file.rb
* test_gem_commands_push_command.rb
* test_gem_commands_owner_command.rb
* test_gem_commands_signin_command.rb:

We should avoid to use copy&paste code.

Reply:
I totally agree with avoid copy-pasting around large chunks of code. I was suggesting to extract all this logic (and the corresponding teardown part if any), and run those methods on the setup callbacks of the specific tests that need it.

This tests probably make direct use of the specific paths being created here, so having the setup more explicit inside these tests makes them easier to follow and understand. And it also avoids all the unnecessary work of creating all these files for every single test file where only 4 of them need that.

If it was most of the tests needing this, then I would understand putting this in the global setup method (which is already a monster by the way), but that's not the case here.

Reply:
I did split them from global setup scope at 7b360d9